Model annotations

NTBiotech opened the pull request:
The model-annotations branch adds annotations to model.py. This includes docstrings, in the google format, as well as comments to explain the code itself. 
There have also been small changes to readablility and pylint conformity, such as proper indentations and the renaming of the Omega_Iterative class to OmegaIterative (conforming to naming conventions).
Please feel free to review and critique the suggested improvements.

NTBiotech committed:
annotations to model.py

NTBiotech committed:
Annotations of Bicycle.__init__()

NTBiotech committed:
Annotation of lyapunov_direct function

NTBiotech committed:
Further annotation and pylint optimization of model.py

NTBiotech committed:
Merge branch 'main' into model-annotations

NTBiotech committed:
Further annotations for functions and classes in data.py, callbacks.py, plotting.py and dictlogger.py

martinrohbeck commented on src/bicycle/utils/training.py:
```diff
@@ -37,6 +37,28 @@ def sylvester_direct(A, B, C):
+    Solves the continuous-time Lyapunov equation AX + XA^T = -C for X using 
+    the direct Kronecker product method.
+
+    Parameters:
```
This should be called `Args:`.

martinrohbeck commented on src/bicycle/utils/training.py:
```diff
@@ -37,6 +37,28 @@ def sylvester_direct(A, B, C):
+                      AX + XA^T = -C for each batch.
+
+    Notes:
+        - This implementation uses the Kronecker product to construct the system of 
```
The 1st and 2nd notes are not helpful. Can be deleted.

martinrohbeck commented on src/bicycle/utils/plotting.py:
```diff
@@ -19,6 +19,9 @@ def plot_training_results(
     callback=True,
     labels=None,
 ):
+    """
```
If it's only a single sentence comment, I'd format it to be a single line.

martinrohbeck commented on src/bicycle/utils/data.py:
```diff
@@ -555,6 +568,24 @@ def get_diagonal_mask(n_genes, device):
+    Deprecated: Currently not initializing beta! Only beta is used in model!
+
+    Args:
+    init_data: iterable, that contains samples, sim_regime, sample_idx, data_category.
```
indentation is missing

martinrohbeck commented on src/bicycle/utils/data.py:
```diff
@@ -555,6 +568,24 @@ def get_diagonal_mask(n_genes, device):
+    """
+    Function to compute initial tensors for the bicycle model.
+
+    Deprecated: Currently not initializing beta! Only beta is used in model!
```
This line is not helpful. Can you rephrase it?

NTBiotech replied:
I removed the line and added a line in Notes

martinrohbeck commented on src/bicycle/utils/data.py:
```diff
@@ -341,6 +348,12 @@ def create_loaders(
+    Function to create dataloaders for training the bicycle model.
+
+    Returns:
+    tuple: (train_loader, validation_loader, test_loader, covariates)
```
indentation is missing

martinrohbeck commented on src/bicycle/utils/data.py:
```diff
@@ -27,6 +27,13 @@ def create_data(
+    Function to create a synthetic single cell RNA-seq dataset with perturbations.
+
+    Returns:
+    tuple: (gt_dyn, intervened_variables, samples, gt_interv, sim_regime, beta)
```
indentation is missing

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1001,7 +1478,7 @@ def predict_perturbation(
         end_time = time.time()
         print(f"Training took {end_time - start_time:.2f} seconds")
 
-    def forward(self):
```
Was there a reason this had to be changed?

NTBiotech replied:
*pylint* gives this output: 
"src/bicycle/model.py:1480:4: W0221: Variadics removed in overriding 'BICYCLE.forward' method (arguments-differ)"

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -897,12 +1371,15 @@ def predict_step(self, batch, dataloader_idx=0):
 
     def predict_means(self, regimes=[]):
+        """
+        Method to predict the mean latent expression of genes.
```
I'd reformat this to be in a single line

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -867,6 +1328,19 @@ def predict_percentages(self, batch):
 
     def predict_step(self, batch, dataloader_idx=0):
+        """
+        Overrides the pyTorch lightning LightningModule predict_step methods called during :meth:`~pytorch_lightning.trainer.trainer.Trainer.predict`.
```
This method does not need a documentation imo.

NTBiotech replied:
Since the implementation here returns a loss instead of a prediction like the Super, I think it is useful to describe the function.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -856,6 +1303,20 @@ def validation_step(self, batch, batch_idx):
 
     def predict_percentages(self, batch):
+        """
+        Predicts percentage probabilities for a given batch of data.
```
Please mention probabilities of what is predicted

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -616,6 +1045,17 @@ def compute_spectral_loss(self, B):
         return spectral_loss
 
     def training_step(self, batch, batch_idx):
+        """
```
I think no documentation is needed for this method.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -526,21 +935,40 @@ def get_mvn_normal(self, B, alphas, sim_regime, sigmas):
+            torch.distributions.MultivariateNormal
+
+        Notes:
+            - Uses the self.use_encoder attribute to determine the method.
```
This line is not helpful

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -488,20 +868,49 @@ def split_samples(self, samples, sim_regime, sample_idx, data_category):
     def get_mvn_normal(self, B, alphas, sim_regime, sigmas):
+        """
+        Compute the latent multivariate normal distribution of gene expression.
+        E.g. solution of lyapunov equation.
```
This only returns the MVN of z, correct?

NTBiotech replied:
Yes and z is  in a steady state distribution according to the lypunov equation, if I understood it correctly. I can rephrase it though.

martinrohbeck replied:
`z` is describing the steady state distribution, but only the variance of this distribution is the solution to the Lyapunov equation (which is also not directly computed in here). That's why it's confusing.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -488,20 +868,49 @@ def split_samples(self, samples, sim_regime, sample_idx, data_category):
         alphas_broadcasted = alphas[sim_regime]
 
         if B.shape[0] == 1:
+            # remove all dimensions==1
```
I don't understand what this means?

NTBiotech replied:
[torch.Tensor.squeeze](https://docs.pytorch.org/docs/stable/generated/torch.squeeze.html) removes all dimensions, that are of length 1

martinrohbeck replied:
Ah, the way to write it like this is a bit confusing to me. The following code line is a simple `.squeeze()`, so this does not add any helpful information. Can be removed.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -488,20 +868,49 @@ def split_samples(self, samples, sim_regime, sample_idx, data_category):
+
+        Notes:
+            - sim_regime is used to rearrange B and alpha through indexing.
+            - x_bar can be later used as the mean for the latent multivariate normal
```
I don't think it's relevant here for what `x_bar` can be used later.

NTBiotech replied:
I put this line in since in the paper x_bar is z_bar instead and this might be confusing without explanation.

martinrohbeck replied:
The synthetic case assumes `z=x` in the paper (see experiments), e.g.. Eq. 4.1.
I'd say it's clear what this function does, based on the first docstring line `x_bar = B^(-1) @ alpha` and imho the docstring should describe the "local" properties of the function, not what might happen in the future to any of its return statements.
But it's not a hard line for me; so, if you prefer to leave it, we can do so.

NTBiotech replied:
I would like to leave it in, just to prevent any confusion from people reading the code for the first time.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -464,6 +824,26 @@ def scale_losses(self, z_kl, l1_loss, loss_spectral=None, loss_lyapunov=None):
 
     def split_samples(self, samples, sim_regime, sample_idx, data_category):
+        """
+        Splits a samples, sim_regime and sample_idx into training, 
```
```suggestion
        Splits samples, sim_regime and sample_idx into training, 
```

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -453,6 +792,27 @@ def lyapunov_rhs(self, sigmas):
+        """
+        Function to scale the loss outputs. 
+
+        Args:
```
Can be removed if not explained (also not necessary).

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -464,6 +824,26 @@ def scale_losses(self, z_kl, l1_loss, loss_spectral=None, loss_lyapunov=None):
+        validation and testing set according to data_category.
+        
+        Args:
+            samples
```
Everything that is not explained should be removed

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -396,39 +707,67 @@ def get_updated_states(self):
+            sigmas: Matrix with dimensions (batch, genes). Sigma represents the effect of
+                the Wiener Process on the transcription of each gene. 
+
+        Returns:
```
Not needed, since the function is only a single line.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -396,39 +707,67 @@ def get_updated_states(self):
+
+        Args:
+            B: matrix with dimensions (batch, genes, genes).
+            Equates to $$I_G - \beta^T$$.
```
Not useful, if I_G is not defined, and beta not present in the method. If you want to keep this, please refer to the paper directly

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -396,39 +707,67 @@ def get_updated_states(self):
+            Equates to $$I_G - \beta^T$$.
+
+        Notes:
+            - omega factorized with sigma_min to avoid vanishing gradient.
```
The reason for the factorization is that we want to obtain a valid covariance matrix --  as the sum of a positive diagonal matrix and a low-rank positive semidefinite matrix

NTBiotech replied:
That makes sense, thanks for the explanation :+1:

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -346,48 +606,99 @@ def configure_optimizers(self):
+        # initializes betas as merge of
+        # beta and beta_p with gt_interv as mask
         if self.n_factors == 0:
+            # in case of interference
```
What exactly does this mean?

NTBiotech replied:
I meant [CRISPR interference](https://de.wikipedia.org/wiki/CRISPRi).

martinrohbeck replied:
Thanks for the clarification!

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -346,48 +606,99 @@ def configure_optimizers(self):
+              interventions.
+            - Diagonal elements of beta matrices are set to zero to inhibit
+              self-regulation.
+            - The function uses PyTorch operations for tensor manipulation,
```
Not relevant or helpful. Can be deleted

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -328,6 +559,35 @@ def __init__(
+        """
+        Configures and returns the optimizer and learning rate scheduler for the model.
+
+        This method selects the optimizer based on the `self.optimizer` attribute and
```
This paragraph can be removed, since it does not add much value

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -328,6 +559,35 @@ def __init__(
+            - The "adamlrs" option includes a learning rate scheduler that reduces the
+              learning rate when the monitored metric stops improving. The monitored
+              metric is currently set to "train_loss".
+            - Ensure that `self.optimizer_kwargs` contains valid arguments for the
```
Please remove.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -328,6 +559,35 @@ def __init__(
+              metric is currently set to "train_loss".
+            - Ensure that `self.optimizer_kwargs` contains valid arguments for the
+              selected optimizer.
+            - The `FIXME` comment indicates that the monitored metric ("train_loss")
```
Please remove.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -294,14 +517,21 @@ def __init__(
+        else:
+            self.T = torch.tensor(T)  # torch.nn.Parameter(torch.tensor(1.0))
+
+        # For benchmarking
```
What does this mean? Please clarify

NTBiotech replied:
These attributes are not used within the model implementation, but can be used for analysis of performance. In the repository they are only used for benchmarking.

martinrohbeck replied:
```suggestion
        # The following assignments are only relevant for model evaluation
```

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -294,14 +517,21 @@ def __init__(
                 if "w_cov_diag" in self.init_tensors:
                     self.w_cov_diag.data = self.init_tensors["w_cov_diag"]
 
+        if learn_T:
```
Was there a particular reason, why this was moved?

NTBiotech replied:
I think this is a more coherent order for the parameter initiation.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -153,28 +349,36 @@ def __init__(
         self.use_encoder = use_encoder
         self.use_latents = use_latents
-        self.n_factors = n_factors
+        self.n_factors = n_factors      # ==0
```
What does this comment say exactly?

NTBiotech replied:
n_factors is set to 0 by default. I will remove it.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -85,7 +221,7 @@ class BICYCLE(pl.LightningModule):
         self,
         lr,
-        gt_interv,
+        gt_interv,  #ground truth of interventions?
```
No comments here, please.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -37,14 +121,60 @@ def __init__(self, x_dim: int, z_dim: int, n_cond: int, act_fn: object = nn.GELU
+    """
+    This class optimizes omega to fit the latent normal distribution of z.
+    It utilizes continuous lyapunov equation and assumes steady state.
+    The class subsets pyTorch_lightnings LightningModule class.
```
Can be removed, not helpful.

NTBiotech replied:
Consider [PEP257](https://peps.python.org/pep-0257/#:~:text=If%20a%20class,its%20own%20behavior).

martinrohbeck replied:
I've literally never seen this, but fair enough (I also believe the more important pieces are the Override/Extend docstrings in the corresponding methods)

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -37,14 +121,60 @@ def __init__(self, x_dim: int, z_dim: int, n_cond: int, act_fn: object = nn.GELU
-    def __init__(self, alpha, beta, sigma, lr=1e-2, sigma_min=1e-3, device="cpu"):
+class OmegaIterative(pl.LightningModule):
+    """
+    This class optimizes omega to fit the latent normal distribution of z.
```
No, it does not optimize omega. It gives a parametrization for omega and computes B.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -37,14 +121,60 @@ def __init__(self, x_dim: int, z_dim: int, n_cond: int, act_fn: object = nn.GELU
+class OmegaIterative(pl.LightningModule):
+    """
+    This class optimizes omega to fit the latent normal distribution of z.
+    It utilizes continuous lyapunov equation and assumes steady state.
```
There is also no Lyapunov eq. utilized in here

NTBiotech replied:
Then why do we use the difference between the rhs and lhs of the lyapunov equation to compute the loss?

martinrohbeck replied:
Sorry, what I meant is that the core idea of this class is to use the parameters to estimate their "quality" as potential candidates to solve the LE. I think we should explain this concept here instead

NTBiotech replied:
I am not sure I can adequately explain it. Could you write a suggestion?

martinrohbeck replied:
Sure, you can just add something like: "During training, instead of directly solving the Lyapunov equation (LE), it computes the difference between LHS and RHS of the LE and updates omega to minimize this difference. Thereby, we can learn to approximate a solution to the LE."

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -21,7 +67,45 @@ def init_weights(m):
+            Processes the input tensor `x` and returns the mean (`mu`) and
+            variance (`variance`) of the latent distribution.
+
+    Example:
```
I think this example can be removed, since it does not provide much information...

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,64 @@
 from bicycle.utils.training import EarlyStopperTorch, lyapunov_direct
 
-
+# not implemented
```
what does this mean? if it's not implemented in the sense of not used, can we remove it?

NTBiotech replied:
At the time I didn't find an implementation, but I now realize, that is used in the Encode class.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,64 @@
+        m (torch.nn.Module): The neural network layer to initialize. This function specifically
+                         targets layers of type nn.Conv2d, nn.Linear, and nn.ConvTranspose2d.
+
+    Behavior:
```
`Behavior` (I've never seen this keyword in a google docstring) and `Note` can be removedd, imo, since it's very repetitive and does not add much value.

martinrohbeck commented on src/bicycle/callbacks.py:
```diff
@@ -65,7 +65,25 @@ def on_validation_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModul
 class GenerateCallback(pl.Callback):
+    """
+    Class for generating a callback at the end of each training epoch of the bicycle model.
+    Subclasses pytorch_lightnings Callback function.
```
This line should be removed, unnecessary detail.

NTBiotech replied:
I can remove it but I would suggest to consider [PEP 257](https://peps.python.org/pep-0257/#:~:text=If%20a%20class,its%20own%20behavior).

martinrohbeck commented on src/bicycle/callbacks.py:
```diff
@@ -65,7 +65,25 @@ def on_validation_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModul
+
+    Notes:
+    - Calls the plot_training_results function from bicycle.utils.plotting
+        after each epoch thats a multiple of self.plot_epoch_callback. 
```
```suggestion
        after each epoch that's a multiple of `plot_epoch_callback`. 
```

martinrohbeck commented on src/bicycle/callbacks.py:
```diff
@@ -65,7 +65,25 @@ def on_validation_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModul
+        Initializes the GenerateCallback function for plotting training results during training.
+
+        Args:
+        file_name_plot: str containing the file_name for the saved plots.
```
The Indentation is missing

martinrohbeck commented on src/bicycle/dictlogger.py:
```diff
@@ -14,6 +14,10 @@ def __missing__(self, key):
 class DictLogger(Logger):
+    """
+    Class for logging model hyperparameters and metrics. 
+    Subclasses the pytorch_lightning.loggers Logger class.
```
This line should be removed, it is an unnecessary detail.

NTBiotech replied:
Same as [before](https://github.com/PMBio/bicycle/pull/4/files#r2084022221). I get that it doesn't add much, but in my opinion mentioning it helps the understanding of the function while working with the code. I would add more detail (like the differences between Super and Subclass), but I am not very knowledgeable in Logger and Callback classes.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,64 @@
+"""
```
I'd remove this comment at the top of the file, since everything is commented in detail below, and it does not add that much value (e.g., it's obvious that this is the code for the Bicycle paper, etc.)

NTBiotech replied:
I think its good practice to describe the module briefly in a docstring. It helps for people trying to implement our model on their own and doesn't really have any downsides. *Pylint* gave me the suggestion for it and I think it adds value.

martinrohbeck replied:
If you say it's helpful, then I agree that we can leave it.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -37,14 +121,60 @@ def __init__(self, x_dim: int, z_dim: int, n_cond: int, act_fn: object = nn.GELU
 
-class Omega_Iterative(pl.LightningModule):
-    def __init__(self, alpha, beta, sigma, lr=1e-2, sigma_min=1e-3, device="cpu"):
+class OmegaIterative(pl.LightningModule):
```
Maybe we should rename the class?

NTBiotech replied:
I think that's a great idea. It is also only used in the predict_perturbation function which seems to be not fully implemented yet.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -37,14 +121,60 @@ def __init__(self, x_dim: int, z_dim: int, n_cond: int, act_fn: object = nn.GELU
         self.net.apply(init_weights)
 
     def forward(self, x):
+        """Overrides pyTorchs nn.Module.forward function."""
```
Can be removed.

martinrohbeck reviewed (changes requested):
Thanks for making the effort to add comments 👍  . I left some feedback (mostly about comments in the code that can be removed, since they don't add much value).

If you are interested in some perspective on how to write good comments, you can find one here: [https://realpython.com/python-comments-guide/#python-commenting-best-practices](https://realpython.com/python-comments-guide/#python-commenting-best-practices). 
I think the key should be that comments have a real value and do not explain what can already be easily derived from the code.

Let me know if you have any questions :]

NTBiotech committed:
Apply suggestions from code review

Co-authored-by: Martin Rohbeck <35061428+martinrohbeck@users.noreply.github.com>

NTBiotech committed:
Compliance with code review from pull request Model annotations #4

Co-authored-by: Martin Rohbeck <35061428+martinrohbeck@users.noreply.github.com>

NTBiotech committed:
Update src/bicycle/model.py

Co-authored-by: Martin Rohbeck <35061428+martinrohbeck@users.noreply.github.com>

NTBiotech committed:
Compliance with code review.

Co-authored-by: Martin Rohbeck <martin.rohbeck@quantco.com>

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,55 @@
+Classes:
+    - Encoder: Encodes input data into a latent space representation for use in a
+      Variational Autoencoder (VAE) or similar architecture.
+    - Omega_Iterative: Optimizes a matrix `omega` using Lyapunov stability principles
```
What are "Lyapunov stability principles"?

NTBiotech replied:
[Lyapunovs criteria](https://en.wikipedia.org/wiki/Lyapunov_stability#Lyapunov's_second_method_for_stability:~:text=Consider%20a%20function,is%20required) for stability of a system, that are applicable for the [time-continuous lyapunov equation](https://en.wikipedia.org/wiki/Lyapunov_equation#:~:text=the%20continuous%2Dtime%20Lyapunov%20equation%20is,) we are using.

martinrohbeck replied:
Then please define the `V(x)` (as used in your reference) in the docstring as well, and explain how we satisfy/leverage the three constraints from in the article.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,55 @@
+Classes:
+    - Encoder: Encodes input data into a latent space representation for use in a
+      Variational Autoencoder (VAE) or similar architecture.
+    - Omega_Iterative: Optimizes a matrix `omega` using Lyapunov stability principles
```
```suggestion
    - OmegaIterative: Optimizes a matrix `omega` using Lyapunov stability principles
```

NTBiotech replied:
I changed the name to comply pylint. It isn't relevant in my opinion though, so I can change it again.

NTBiotech replied:
You can also check PEP8 for [naming conventions](https://peps.python.org/pep-0008/#comments:~:text=The%20naming%20convention%20for%20functions%20may%20be%20used%20instead%20in%20cases%20where%20the%20interface%20is%20documented%20and%20used%20primarily%20as%20a%20callable.). But if there is a problem with downstream compatability, I can change it again.

martinrohbeck replied:
No, I think your change is totally fine, we just need to be consistent in the files.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,55 @@
+    - Implements spectral loss to ensure stability in gene-gene interaction matrices.
+    - Supports early stopping during training for efficient convergence.
+
+Dependencies:
```
I guess this part belongs into the env file and can be removed here?

NTBiotech replied:
Its already in the env file. I can remove it.

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,18 +1,55 @@
+      Lyapunov loss, and sparsity loss.
+    - Handles various intervention types, including CRISPR-based perturbations (e.g., Cas9, dCas9).
+    - Provides options for latent variable modeling using an encoder or direct parameterization.
+    - Implements spectral loss to ensure stability in gene-gene interaction matrices.
```
Can this be combined with the first bullet point (which already addresses the losses)?

NTBiotech committed:
Compliance with PR #4.

Co-authored-by: Martin Rohbeck <martin.rohbeck@quantco.com>

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -43,8 +107,53 @@ def forward(self, x):
+            used in Lyapunov computations.
+
+    Methods:
+        configure_optimizers():
```
Why are these docstrings here instead of directly with the methods below?
Also, the configure_optimizers can be removed, because it's only a single, obvious line of code.

martinrohbeck commented:
I think we are already close to the finish line :]

martinrohbeck commented:
If you address the remaining comments, we can merge this.

NTBiotech committed:
Compliance with PR #4.

Co-authored-by: Martin Rohbeck <martin.rohbeck@quantco.com>

NTBiotech commented:
I think I have addressed the comments, that were made up to this point. If there is anything else, please feel free to bring it up. Thanks for the feedback!

martinrohbeck committed:
minor docstring changes to callbacks.py

martinrohbeck committed:
minor docstring changes to data.py

martinrohbeck committed:
Minor docstring edits for plotting.py

martinrohbeck committed:
Minor docstring edits for training.py

martinrohbeck committed:
Update model.py

Minor docstring updates to model.py

martinrohbeck commented on src/bicycle/model.py:
```diff
@@ -1,27 +1,95 @@
+        1. V(x)=0 if x=0
+        2. V(x)>0 if x!=0
+        3. V°(x) = dV(t)/dt = sum(dV/dx_i * f_i(x)) = nabla V *f(x) =< 0 for all x!=0
+        The model assumes V(x) as a time-continuous Lyapunov equation:
```
Can you explain/clarify
(1) where the three properties are used/satisified in the code?
(2) fix the V(x) with the correct parameters, since there is no x in the Lyapunov equation...

NTBiotech replied:
I can change to notation. But there is no direct reference to Lyapunovs criteria in the code itself. They are assumptions that are made for the model.

martinrohbeck commented:
Thanks again! I made a few more minor changes to the docstrings. Could you address the last comment I left, and then feel free to press the merge button :]!

NTBiotech committed:
Compliance with PR #4

Co-authored-by: Martin Rohbeck <martin.rohbeck@quantco.com>

martinrohbeck committed:
Removed comments
